project3/coordinator/messenger.h

```diff
@@ -0,0 +1,77 @@
+/**
+ * @File Messenger to be utilized by the coordinator.
+ */
+#ifndef PROJECT3_MESSENGER_H
+#define PROJECT3_MESSENGER_H
+
+#include <pub_sub_messages.pb.h>
+#include <sstream>
+#include <string>
+#include <memory>
+
+#include "connected_participants.h"
+#include "message_log.h"
+#include "pub_sub_messages.pb.h"
+#include "queue/queue.h"
+#include "wire_protocol/wire_protocol.h"
+namespace coordinator::messenger {
+using namespace coordinator::message_log;
+using namespace coordinator::connected_participants;
+/**
+ * @class Handles message broadcasting to participants.
+ *
+ */
+class Messenger {
```

Comment:
I would be tempted to say that the `Messenger` class should only be responsible for sending to a single participant, or it should only be responsible for sending to a group of them. Here you try to split the difference, having both a multicast group and a single participant that this `Messenger` is specially in-charge of. Perhaps this should be split up into two classes? Or at the very least, maybe make `SendMissedMessages` participant-agnostic, i.e. allow the receiving participant to be specified as an argument.

Reply:
Actually, now I'm rethinking this suggestion a little. It seems that the way you've intended for this class to be used is to encapsulate communication with a *single* participant. To that end, the best idea is probably to make this class responsible only for sending to a single `participant`, and have a separate class that aggregates multiple `Messengers` for handling broadcasting. If you do this right, it would also solve your broadcasting concurrency issue.

Reply:
Yup, this is exactly what I was thinking of.